q2_cutadapt/_demux.py

```diff
@@ -33,29 +34,61 @@ def run_command(cmd, verbose=True):
 
 
 def _build_demux_command(seqs_dir_fmt, barcode_fasta, per_sample_dir_fmt,
-                         untrimmed_dir_fmt, error_tolerance):
+                         untrimmed_dir_fmt, error_rate):
     cmd = ['cutadapt',
-           '-g', 'file:%s' % barcode_fasta.name,
-           '-e', str(error_tolerance),
-           # {name} is a cutadapt convention for interpolating the sample id
-           # into the filename.
-           '-o', '%s/{name}.fastq.gz' % str(per_sample_dir_fmt),
-           '--untrimmed-output',
-           '%s/forward.fastq.gz' % str(untrimmed_dir_fmt),
-           str(seqs_dir_fmt.file.view(FastqGzFormat)),
+           '--front', 'file:%s' % barcode_fasta.name,
+           '--error-rate', str(error_rate),
            ]
+    # PAIRED-END
+    if isinstance(seqs_dir_fmt, MultiplexedPairedEndBarcodeInSequenceDirFmt):
+        cmd = cmd + [
+            # {name} is a cutadapt convention for interpolating the sample id
+            # into the filename.
+            '-o', os.path.join(str(per_sample_dir_fmt), '{name}.1.fastq.gz'),
+            '-p', os.path.join(str(per_sample_dir_fmt), '{name}.2.fastq.gz'),
+            '--untrimmed-output',
+            os.path.join(str(untrimmed_dir_fmt), 'forward.fastq.gz'),
+            '--untrimmed-paired-output',
+            os.path.join(str(untrimmed_dir_fmt), 'reverse.fastq.gz'),
+            str(seqs_dir_fmt.forward_sequences.view(FastqGzFormat)),
+            str(seqs_dir_fmt.reverse_sequences.view(FastqGzFormat)),
+            ]
+    # SINGLE-END
+    else:
+        cmd = cmd + [
+            # {name} is a cutadapt convention for interpolating the sample id
+            # into the filename.
+            '-o', os.path.join(str(per_sample_dir_fmt), '{name}.fastq.gz'),
+            '--untrimmed-output',
+            os.path.join(str(untrimmed_dir_fmt), 'forward.fastq.gz'),
+            str(seqs_dir_fmt.file.view(FastqGzFormat)),
+            ]
     return cmd
 
 
-def _rename_files(per_sample_dir_fmt, barcode_series):
-    for (sample_id, barcode_id) in barcode_series.iteritems():
-        out_fp = per_sample_dir_fmt.sequences.path_maker(sample_id=sample_id,
-                                                         barcode_id=barcode_id,
-                                                         lane_number=1,
-                                                         read_number=1)
-        src = os.path.join(str(per_sample_dir_fmt), '%s.fastq.gz' % sample_id)
-        if os.path.isfile(src):
-            os.rename(src, str(out_fp))
+def _rename_files(seqs_dir_fmt, per_sample_dir_fmt, barcode_series):
+    # PAIRED-END
+    if isinstance(seqs_dir_fmt, MultiplexedPairedEndBarcodeInSequenceDirFmt):
```

Comment:
You could factor out this if-else branch to something like this:
```python
read_directions = [1]
if isinstance(...)
    read_directions.append(2)

for (sample_id, barcode_id) in barcode_series.iteritems():
    for read_direction in read_directions:
        out_fp = ...
        src = ...
        if os.path.isfile(src):
            os.rename(...)
```
Then the business logic only has to exist once

Reply:
This only works if the initial filename is consistent between single and paired-end.